core/internal/pathtree/pathtree.go

```diff
@@ -86,54 +114,125 @@ func (pathTree *PathTree) ApplyUpdate(
 			continue
 		}
 
-		if err := updateAtPath(pathTree.tree, item.Path, value); err != nil {
+		if err := updateAtPath(pt.tree, item.Path, value); err != nil {
 			onError(err)
 			continue
 		}
 	}
 }
 
-func (pathTree *PathTree) ApplyRemove(
+// Removes values from the tree.
+func (pt *PathTree) ApplyRemove(
 	items []*PathItem,
 	onError func(error),
 ) {
 	for _, item := range items {
-		pathTree.removeAtPath(item.Path)
+		pt.removeAtPath(item.Path)
+	}
+}
+
+// Removes the value at the path in the config tree.
+func (pt *PathTree) removeAtPath(path TreePath) {
+	prefix := path[:len(path)-1]
+	key := path[len(path)-1]
+
+	subtree := getSubtree(pt.tree, prefix)
+	if subtree != nil {
+		delete(subtree, key)
+	}
+}
+
+// Uses the given subtree for keys that aren't already set.
+func (pt *PathTree) AddUnsetKeysFromSubtree(
+	tree TreeData,
+	path TreePath,
+) error {
+	oldSubtree := getSubtree(tree, path)
+	if oldSubtree == nil {
+		return nil
 	}
+
+	newSubtree, err := getOrMakeSubtree(pt.tree, path)
+	if err != nil {
+		return err
+	}
+
+	for key, value := range oldSubtree {
+		if _, exists := newSubtree[key]; !exists {
+			newSubtree[key] = value
+		}
+	}
+
+	return nil
 }
 
 // Serializes the object to send to the backend.
-func (pathTree *PathTree) Serialize(format Format, processValue func(any) any) ([]byte, error) {
+func (pt *PathTree) Serialize(
+	format Format,
+	processValue func(any) any,
+) ([]byte, error) {
 	// A configuration dict in the format expected by the backend.
-	serialized := make(map[string]any)
-	for treeKey, treeValue := range pathTree.tree {
+	value := make(map[string]any)
+	for treeKey, treeValue := range pt.tree {
 		if processValue == nil {
-			serialized[treeKey] = treeValue
+			value[treeKey] = treeValue
 		} else {
-			serialized[treeKey] = processValue(treeValue)
+			value[treeKey] = processValue(treeValue)
 		}
 	}
+	return marshal(value, format)
+}
 
-	switch format {
-	case FormatYaml:
-		return yaml.Marshal(serialized)
-	case FormatJson:
-		return json.Marshal(serialized)
+// FlattenAndSerialize flattens the tree into a slice of leaves and marshals the values.
```

Comment:
similar to the other comment: PathTree shouldn't handle JSON. This function can then just be `Flatten` and return whatever leaf values were added to the tree by the higher level code.

Reply:
i'm ok to take it out as it iterates over the values again anyways